For this project I wanted to use a classifier to predict a players position based on height and weight. I feel like certain players fall in certain positional groups based on height and weight and I wanted to see how accurately we could predict this using only those two characteristics. I got a dataset from Kaggle to do this.

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv(r'C:\Users\michael.samson\Downloads\Basic_Stats.csv.zip',compression='zip')
df.head()

Age        Birth Place    Birthday           College Current Status  \
0   NaN  Grand Rapids , MI   5/23/1921        Notre Dame        Retired   
1   NaN        Dayton , OH  12/21/1930            Dayton        Retired   
2  56.0        Temple , TX   9/11/1960    Louisiana Tech        Retired   
3  30.0   New Orleans , LA   9/30/1986               LSU        Retired   
4  25.0       Detroit , MI   3/31/1992  Central Michigan         Active   

          Current Team  Experience  Height (inches)     High School  \
0                  NaN   3 Seasons             71.0             NaN   
1                  NaN    1 Season             70.0             NaN   
2                  NaN    1 Season             74.0             NaN   
3                  NaN   5 Seasons             73.0             NaN   
4  Pittsburgh Steelers  3rd season             77.0  Clintondale HS   

          High School Location            Name  Number            Player Id  \
0                          NaN     Evans, Fred     NaN    fredevans/2513736   
1                          NaN      Raiff, Jim     NaN     jimraiff/2523700   
2                          NaN   Fowler, Bobby     NaN  bobbyfowler/2514295   
3                          NaN  Johnson, Quinn     NaN   quinnjohnson/79593   
4  Clinton Twp.,Macomb Co., MI    Walton, L.T.    96.0   l.t.walton/2552444   

  Position  Weight (lbs) Years Played  
0      NaN         185.0  1946 - 1948  
1      NaN         235.0  1954 - 1954  
2      NaN         230.0  1985 - 1985  
3      NaN         255.0  2009 - 2013  
4       DE         305.0          NaN

I wanted to clean this data by only getting the columns I wanted and also dropping all positional NA's from the dataset

In [3]:
new_data = df.iloc[:,[7,13,14]]
new_data = new_data[new_data['Position'].notna()]
new_data

Height (inches) Position  Weight (lbs)
4                 77.0       DE         305.0
7                 77.0       TE         250.0
17                74.0       TE         220.0
35                71.0       SS         200.0
37                72.0       SS         205.0
...                ...      ...           ...
17127             73.0       DB         210.0
17140             79.0       DE         287.0
17151             75.0      OLB         255.0
17155             80.0        T         320.0
17157             72.0       WR         205.0

[3096 rows x 3 columns]

Now we need to import the sklearn library that has the random forest classifier and we need to train our model. Our predictive values here are height and weight and the (dependent) variable we want to predict is position 

In [4]:
# Import train_test_split function
from sklearn.model_selection import train_test_split

X=new_data[['Height (inches)','Weight (lbs)']]  # Features
y=new_data['Position']  # Labels

# Split dataset into training set and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

In [5]:
#Import Random Forest Model
from sklearn.ensemble import RandomForestClassifier

#Create a Gaussian Classifier
clf=RandomForestClassifier(n_estimators=100,random_state = 42)

#Train the model using the training sets y_pred=clf.predict(X_test)
clf.fit(X_train,y_train)

y_pred=clf.predict(X_test)

In [6]:
#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics
# Model Accuracy, how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.3078579117330463


Overall our accuracy is very poor. If we clean our data for all offensive positions will our accuracy increase? Let's check.

In [7]:
new_data.Position.unique()

array(['DE', 'TE', 'SS', 'DB', 'OT', 'CB', 'G', 'OG', 'C', 'QB', 'FB',
       'MLB', 'FS', 'DT', 'WR', 'OLB', 'T', 'LB', 'RB', 'ILB', 'K', 'NT',
       'LS', 'P', 'SAF', 'DL', 'OL'], dtype=object)

In [8]:
new_values = ['TE', 'QB', 'FB', 'WR', 'RB', 'K', 'P', 'OL'] #only filtering for O line, running back, wide receiver, qb and special teams

offense_only = new_data.loc[new_data['Position'].isin(new_values)]
offense_only.head()

Height (inches) Position  Weight (lbs)
7               77.0       TE         250.0
17              74.0       TE         220.0
72              76.0       QB         230.0
79              70.0       FB         259.0
139             73.0       WR         195.0

In [9]:
# Import train_test_split function
from sklearn.model_selection import train_test_split

X=offense_only[['Height (inches)','Weight (lbs)']]  # Features
y=offense_only['Position']  # Labels

# Split dataset into training set and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

In [10]:
#Import Random Forest Model
from sklearn.ensemble import RandomForestClassifier

#Create a Gaussian Classifier
clf=RandomForestClassifier(n_estimators=100,random_state = 42)

#Train the model using the training sets y_pred=clf.predict(X_test)
clf.fit(X_train,y_train)

y_pred=clf.predict(X_test)

In [11]:
#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics
# Model Accuracy, how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.6492307692307693


Here we see by selecting more specific data the accuracy now is way higher, actually predicting the correct position 65% of the time